sql: support `"{}"` format for array column in COPY FROM STDIN WITH CSV

ZhouXing19 opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/70728

In postgreSQL, user can use the `"{}"` format
to insert an array column. Currently, Cockroach does not
supports this format, but it allows the `"ARRAY[]"`
format.

This commit adds support for the `"{}"` format with
`COPY FROM STDIN WITH CSV` syntax.

Usage:

```
movr=> CREATE TABLE c (d INT ARRAY);
CREATE TABLE

movr=> COPY c(d) FROM STDIN WITH CSV;
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> "{1,2}"
>> "{3,4,5}"
>> \.
COPY 2
movr=> SELECT * FROM c;
   d
-------
 {1,2}
 {3,4,5}
(2 rows)
```

Release note (bug fix): support `"{}"` format for array column in COPY FROM STDIN WITH CSV
Release justification: None

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71688)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @ZhouXing19)

___
*[pkg/sql/rowenc/roundtrip_format.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71688#-MmOqyAKDLg9C3BqPR_W:-MmOqyAKDLg9C3BqPR_X:bumf1f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d7e0f9be0e6c2af1898111c0a0928c5e92837b2/pkg/sql/rowenc/roundtrip_format.go#L68)):*
> ```Go
> 
> // formatArrayString is to convert an array with format `{}` to fotmat `ARRAY[]`.
> func formatArrayString(s string) string {
> ```

nice tests! the concern i have is that this function seems like it might be a bit brittle. i'm not sure if it will handle special characters like `"`, `,`, or `]` (and others) correctly.

luckily, we already do have a way of parsing arrays with the `ParseAndRequireString` function

i'm curious what happens if we change this line: https://github.com/cockroachdb/cockroach/blob/97c3d0bd94fbc5a6f52f87836976f1c32183b790/pkg/sql/copy.go#L467

so it calls `tree.ParseAndRequireString` instead?


it seems like that might make it so that COPY doesn't handle the `ARRAY[...]` syntax. but i'm not sure if it needs to. does postgres handle `ARRAY[...]`?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71688)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/rowenc/roundtrip_format.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71688#-MmOqyAKDLg9C3BqPR_W:-MmPHzuM2oVqC3qHMg3M:b-xun7vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d7e0f9be0e6c2af1898111c0a0928c5e92837b2/pkg/sql/rowenc/roundtrip_format.go#L68)):*
1. Yes I have the same concern -- is a string with `{}` format always a valid array? Will look into more edge cases.

2. Actually, `ParseAndRequireString` is called inside of `ParseDatumStringAs`. Here's [a comment](https://github.com/cockroachdb/cockroach/blob/9d7e0f9be0e6c2af1898111c0a0928c5e92837b2/pkg/sql/rowenc/roundtrip_format.go#L26-L27) in `ParseDatumStringAs`:

> 	// We use a different parser for array types because ParseAndRequireString only parses
>	// the internal postgres string representation of arrays.

But unfortunately, the `parseAsTyp` inside of  `ParseDatumStringAs` doesn't support an input string with `{}` format, because the parser can't parse the `SET ROW ( { expr } )` syntax. 

Actually, I think to solve this issue, we can choose one of the following:
  1)  add parser support for `SET ROW ( { expr } )`, keep everything unchanged in `copy.go`. could be a bit more complicated;
  2)  (the current one) convert the `{}` array to `ARRAY[]` in `copy.go`.

Any idea which one we should take?

3. Postgre doesn't support the `"ARRAY[]"` in COPY (but they do in INSERT):

```
postgres=# COPY c(d) FROM STDIN WITH CSV;
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> "ARRAY[1,2]"
>> \.
ERROR:  malformed array literal: "ARRAY[1,2]"
DETAIL:  Array value must start with "{" or dimension information.
```


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql: support "{}" format for array column in COPY FROM STDIN WITH CSV

In postgreSQL, user can use the `"{}"` format
to insert an array column. Currently, Cockroach does not
supports this format, but it allows the `"ARRAY[]"`
format.

This commit adds support for the `"{}"` format with
`COPY FROM STDIN WITH CSV` syntax.

Usage:

```
movr=> CREATE TABLE c (d INT ARRAY);
CREATE TABLE

movr=> COPY c(d) FROM STDIN WITH CSV;
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> "{1,2}"
>> "{3,4,5}"
>> \.
COPY 2
movr=> SELECT * FROM c;
   d
-------
 {1,2}
 {3,4,5}
(2 rows)
```

Fixes #70728
Release note (bug fix): support `"{}"` format for array column in COPY FROM STDIN
WITH CSV

rafiss reviewed (approved):
very nice! thanks for diving so deep to fix this

rafiss commented:
nit: on the commit message, I think we should call this a `Release not (bug fix):`

ZhouXing19 commented:
TFTR! bors r=rafiss

rafiss commented:
tip: the bors comment needs to be on its own line :)

bors r+
